python/src/iceberg/types.py

```diff
@@ -43,6 +46,11 @@ def __new__(cls, *args, **kwargs):
         return cls._instance
 
 
+class Truncatable:
+    def truncate(self, value, width: int):
+        raise ValueError(f"Cannot truncate type: {self}")
```

Comment:
Is the idea here that if a type inherits from `Truncatable` but doesn't have a truncate method, then this generic `ValueError` is raised? If the type is not truncatable, why would it inherit from this class at all? It seems like a level of indirection that's not required.

What do you think of just leaving this class out and including a `truncate` method for the Type classes that support it? If there's somewhere we need to check if a type is truncatable, a try/except that catches an `AttributeError` that's raised for free should be just as performant as an `isinstance(type, Truncatable)` and might even be more pythonic.

Reply:
@samredai  do you mean to just `pass` instead of `raise` error here?

The good thing with `Truncatable` is that we can use it in the type hint. But I am OK to go with the traditional try/except approach for this.

Reply:
Ok I see now the idea is to use the class for type annotations. In that case I'm just wondering how we would ever hit this `ValueError` if there are no mypy typecheck failures then plus the fact that everything that inherits from `Truncatable` will indeed override the truncate method.

I think the goal is to communicate that this is an abc class and an abstract method, maybe we can do that more formally?
```py
from abc import ABC, abstractmethod

class Truncatable(ABC):
    @abstractmethod
    def truncate(self, value, width: int):
        ...
```
The added benefit here is that if for some reason a class does not override the `truncate` method, that will fail immediately during init:
```py
class Foo(Truncatable):
    pass

Foo()  # TypeError: Can't instantiate abstract class Foo with abstract method truncate
```
instead of waiting until the `truncate` method is accessed later on to raise the `ValueError`.

Reply:
Yep, I thought about that. I did not use it to keep the code simple. I will update the code to use ABC class here.